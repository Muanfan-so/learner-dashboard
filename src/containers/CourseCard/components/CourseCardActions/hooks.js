import { handleEvent } from 'data/services/segment/utils';
import { eventNames } from 'data/services/segment/constants';

export const useCourseCardActionData = () => {
  const trackUpgradeClick = () => {
    handleEvent(eventNames.upgradeCourse, {
      pageName: 'learner_home',
      linkType: 'button',
      linkCategory: 'green_upgrade',
    });
  };

  return {
    trackUpgradeClick,
  };
};

export default useCourseCardActionData;
