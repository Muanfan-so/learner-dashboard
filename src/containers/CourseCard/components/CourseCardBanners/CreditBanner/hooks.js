import { StrictDict } from 'utils';
import { hooks as appHooks } from 'data/redux';

import ApprovedContent from './views/ApprovedContent';
import EligibleContent from './views/EligibleContent';
import MustRequestContent from './views/MustRequestContent';
import PendingContent from './views/PendingContent';
import RejectedContent from './views/RejectedContent';

export const statusComponents = StrictDict({
  pending: PendingContent,
  approved: ApprovedContent,
  rejected: RejectedContent,
});

export const useCreditBannerData = (cardId) => {
  const credit = appHooks.useCardCreditData(cardId);
  const { supportEmail } = appHooks.usePlatformSettingsData();
  if (!credit.isEligible) { return null; }

  const { error, purchased, requestStatus } = credit;
  let ContentComponent = EligibleContent;
  if (purchased) {
    if (requestStatus == null) {
      ContentComponent = MustRequestContent;
    } else if (Object.keys(statusComponents).includes(requestStatus)) {
      ContentComponent = statusComponents[requestStatus];
    }
    // Current behavior is to show Elligible State if unknown request status is returned
  }
  return {
    ContentComponent,
    error,
    supportEmail,
  };
};

export default {
  useCreditBannerData,
};
