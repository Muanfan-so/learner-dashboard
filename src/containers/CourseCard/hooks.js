import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

export const useCardData = ({ cardId }) => {
  const { formatMessage } = useIntl();
  const { title, bannerImgSrc } = appHooks.useCardCourseData(cardId);
  const { isEnrolled } = appHooks.useCardEnrollmentData(cardId);

  return {
    isEnrolled,
    title,
    bannerImgSrc,
    formatMessage,
  };
};

export default useCardData;
