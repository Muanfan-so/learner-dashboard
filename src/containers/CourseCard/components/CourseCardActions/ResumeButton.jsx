import React from 'react';
import PropTypes from 'prop-types';

import { useIntl } from '@edx/frontend-platform/i18n';

import track from 'tracking';
import { reduxHooks } from 'hooks';
import ActionButton from './ActionButton';
import messages from './messages';

export const ResumeButton = ({ cardId }) => {
  const { resumeUrl } = reduxHooks.useCardCourseRunData(cardId);
  const { hasAccess, isAudit, isAuditAccessExpired } = reduxHooks.useCardEnrollmentData(cardId);
  const { isMasquerading } = reduxHooks.useMasqueradeData();
  const { formatMessage } = useIntl();
  const handleClick = reduxHooks.useTrackCourseEvent(
    track.course.enterCourseClicked,
    cardId,
    resumeUrl,
  );
  return (
    <ActionButton
      disabled={isMasquerading || !hasAccess || (isAudit && isAuditAccessExpired)}
      as="a"
      href="#"
      onClick={handleClick}
    >
      {formatMessage(messages.resume)}
    </ActionButton>
  );
};
ResumeButton.propTypes = {
  cardId: PropTypes.string.isRequired,
};
export default ResumeButton;
