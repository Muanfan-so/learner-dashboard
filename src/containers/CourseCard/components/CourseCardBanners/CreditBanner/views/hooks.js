import React from 'react';

import { StrictDict } from 'utils';
import { apiHooks } from 'hooks';

import * as module from './hooks';

export const state = StrictDict({
  creditRequestData: (val) => React.useState(val), // eslint-disable-line
});

export const useCreditRequestData = (cardId) => {
  const [requestData, setRequestData] = module.state.creditRequestData(null);
  const createCreditApiRequest = apiHooks.useCreateCreditRequest(cardId);
  const createCreditRequest = (e) => {
    e.preventDefault();
    createCreditApiRequest().then(setRequestData);
  };
  return { requestData, createCreditRequest };
};

export default {
  useCreditRequestData,
};
