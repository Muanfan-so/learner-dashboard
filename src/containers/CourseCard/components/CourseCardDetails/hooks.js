import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

import * as hooks from './hooks';
import messages from './messages';

export const useAccessMessage = ({ cardId }) => {
  const { formatMessage, formatDate } = useIntl();
  const enrollment = appHooks.useCardEnrollmentData(cardId);
  const courseRun = appHooks.useCardCourseRunData(cardId);
  if (!courseRun.isStarted) {
    if (!courseRun.startDate) { return null; }
    const startDate = formatDate(courseRun.startDate);
    return formatMessage(messages.courseStarts, { startDate });
  }
  if (enrollment.isEnrolled) {
    const { isArchived, endDate } = courseRun;
    const {
      accessExpirationDate,
      isAuditAccessExpired,
    } = enrollment;

    if (enrollment.isAudit && accessExpirationDate) {
      return formatMessage(
        isAuditAccessExpired ? messages.accessExpired : messages.accessExpires,
        { accessExpirationDate: formatDate(accessExpirationDate) },
      );
    }
    if (!endDate) { return null; }
    return formatMessage(
      isArchived ? messages.courseEnded : messages.courseEnds,
      { endDate: formatDate(endDate) },
    );
  }
  return null;
};

export const useCardDetailsData = ({ dispatch, cardId }) => {
  const { formatMessage } = useIntl();
  const providerName = appHooks.useCardProviderData(cardId).name;
  const { courseNumber } = appHooks.useCardCourseData(cardId);
  const {
    isEntitlement,
    isFulfilled,
    canChange,
  } = appHooks.useCardEntitlementData(cardId);

  const openSessionModal = appHooks.useUpdateSelectSessionModalCallback(dispatch, cardId);

  return {
    providerName: providerName || formatMessage(messages.unknownProviderName),
    accessMessage: hooks.useAccessMessage({ cardId }),
    isEntitlement,
    isFulfilled,
    canChange,
    openSessionModal,
    courseNumber,
    changeOrLeaveSessionMessage: formatMessage(messages.changeOrLeaveSessionButton),
  };
};

export default useCardDetailsData;
