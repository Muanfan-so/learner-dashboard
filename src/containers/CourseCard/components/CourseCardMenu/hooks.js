import React from 'react';
import { StrictDict } from 'utils';

import track from 'tracking';
import { reduxHooks } from 'hooks';

import * as module from './hooks';

export const state = StrictDict({
  isUnenrollConfirmVisible: (val) => React.useState(val), // eslint-disable-line
  isEmailSettingsVisible: (val) => React.useState(val), // eslint-disable-line
});

export const useUnenrollData = () => {
  const [isVisible, setIsVisible] = module.state.isUnenrollConfirmVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useEmailSettings = () => {
  const [isVisible, setIsVisible] = module.state.isEmailSettingsVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useHandleToggleDropdown = (cardId) => {
  const eventName = track.course.courseOptionsDropdownClicked;
  const trackCourseEvent = reduxHooks.useTrackCourseEvent(eventName, cardId);
  return (isOpen) => {
    if (isOpen) { trackCourseEvent(); }
  };
};

export const useCourseCardMenu = (cardId) => {
  const { courseName } = reduxHooks.useCardCourseData(cardId);
  const { isEnrolled, isEmailEnabled } = reduxHooks.useCardEnrollmentData(cardId);
  const { twitter, facebook } = reduxHooks.useCardSocialSettingsData(cardId);
  const { isMasquerading } = reduxHooks.useMasqueradeData();
  const { isEarned } = reduxHooks.useCardCertificateData(cardId);
  const handleTwitterShare = reduxHooks.useTrackCourseEvent(
    track.socialShare,
    cardId,
    'twitter',
  );
  const handleFacebookShare = reduxHooks.useTrackCourseEvent(
    track.socialShare,
    cardId,
    'facebook',
  );

  const showUnenrollItem = isEnrolled && !isEarned;
  const showDropdown = showUnenrollItem || isEmailEnabled || facebook.isEnabled || twitter.isEnabled;

  return {
    courseName,
    isMasquerading,
    isEmailEnabled,
    showUnenrollItem,
    showDropdown,
    facebook,
    twitter,
    handleTwitterShare,
    handleFacebookShare,
  };
};
