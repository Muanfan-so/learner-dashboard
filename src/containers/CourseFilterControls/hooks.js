import React from 'react';
import { useToggle } from '@edx/paragon';

import { StrictDict } from 'utils';

import * as module from './hooks';

export const state = StrictDict({
  target: (val) => React.useState(val), // eslint-disable-line
});

export const useCourseFilterControlsData = ({
  setFilters,
  setSortBy,
}) => {
  const [isOpen, open, close] = useToggle(false);
  const [target, setTarget] = module.state.target(null);
  const handleFilterChange = ({ target: { checked, value } }) => {
    const update = checked ? setFilters.add : setFilters.remove;
    update(value);
  };
  const handleSortChange = ({ target: { value } }) => {
    setSortBy(value);
  };
  return {
    isOpen,
    open,
    close,
    target,
    setTarget,
    handleFilterChange,
    handleSortChange,
  };
};

export default useCourseFilterControlsData;
