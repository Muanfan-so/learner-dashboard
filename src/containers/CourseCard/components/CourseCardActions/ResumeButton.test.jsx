import { shallow } from 'enzyme';

import { htmlProps } from 'data/constants/htmlKeys';
import { reduxHooks } from 'hooks';
import track from 'tracking';
import ResumeButton from './ResumeButton';

jest.mock('hooks', () => ({
  reduxHooks: {
    useCardCourseRunData: jest.fn(() => ({ resumeUrl: 'resumeUrl' })),
    useCardEnrollmentData: jest.fn(() => ({
      hasAccess: true,
      isAudit: true,
      isAuditAccessExpired: false,
    })),
    useMasqueradeData: jest.fn(() => ({ isMasquerading: false })),
    useTrackCourseEvent: (eventName, cardId, url) => jest
      .fn()
      .mockName(`useTrackCourseEvent('${eventName}', '${cardId}', '${url}')`),
  },
}));
jest.mock('tracking', () => ({
  course: {
    enterCourseClicked: 'enterCourseClicked',
  },
}));
jest.mock('./ActionButton', () => 'ActionButton');

const { resumeUrl } = reduxHooks.useCardCourseRunData();

describe('ResumeButton', () => {
  const props = {
    cardId: 'cardId',
  };
  describe('snapshot', () => {
    test('renders default button when learner has access to the course', () => {
      const wrapper = shallow(<ResumeButton {...props} />);
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.prop(htmlProps.disabled)).toEqual(false);
      expect(wrapper.prop(htmlProps.onClick).getMockName()).toContain(
        'useTrackCourseEvent',
        track.course.enterCourseClicked,
        props.cardId,
        resumeUrl,
      );
    });
  });
  describe('behavior', () => {
    it('initializes course run data based on cardId', () => {
      shallow(<ResumeButton {...props} />);
      expect(reduxHooks.useCardCourseRunData).toHaveBeenCalledWith(props.cardId);
    });
    it('initializes course enrollment data based on cardId', () => {
      shallow(<ResumeButton {...props} />);
      expect(reduxHooks.useCardEnrollmentData).toHaveBeenCalledWith(props.cardId);
    });
    describe('disabled states', () => {
      test('masquerading', () => {
        reduxHooks.useMasqueradeData.mockReturnValueOnce({ isMasquerading: true });
        const wrapper = shallow(<ResumeButton {...props} />);
        expect(wrapper.prop(htmlProps.disabled)).toEqual(true);
      });
      test('learner does not have access', () => {
        reduxHooks.useCardEnrollmentData.mockReturnValueOnce({
          hasAccess: false,
          isAudit: true,
          isAuditAccessExpired: false,
        });
        const wrapper = shallow(<ResumeButton {...props} />);
        expect(wrapper.prop(htmlProps.disabled)).toEqual(true);
      });
      test('audit access expired', () => {
        reduxHooks.useCardEnrollmentData.mockReturnValueOnce({
          hasAccess: true,
          isAudit: true,
          isAuditAccessExpired: true,
        });
        const wrapper = shallow(<ResumeButton {...props} />);
        expect(wrapper.prop(htmlProps.disabled)).toEqual(true);
      });
    });
  });
});
