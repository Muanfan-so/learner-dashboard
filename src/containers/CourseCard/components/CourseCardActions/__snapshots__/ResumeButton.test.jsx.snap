// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResumeButton snapshot renders default button when learner has access to the course 1`] = `
<Button
  as="a"
  disabled={false}
  href="resumeUrl"
>
  Resume
</Button>
`;
